Fixes swallowed exceptions in computation thread

sylvlecl opened the pull request:
See #1681 

Exceptions are now transferred from the computation thread to the main one using std::future mechanisms.
Only the first encountered exception is thrown in the main thread.

**Note**
As a side addition, the errors occurring during the writing of results to zip now cause exceptions that are also "transferred" to the main thread, when calling `IResultWriter::flush()`.

sylvlecl committed:
Fixes swallowed exceptions in computation thread

Exceptions are now transferred from the computation thread
to the main one using std::future mechanisms.
Only the first encountered exception is thrown in the main
thread.

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

flomnes commented on src/tests/src/libs/antares/concurrency/test_concurrency.cpp:
```diff
@@ -0,0 +1,76 @@
+}
+
+class TestException1 {};
+class TestException2 {};
```
```suggestion
template<int n>
class TestException {};
```

sylvlecl replied:
Done

flomnes commented on src/tests/src/libs/antares/concurrency/test_concurrency.cpp:
```diff
@@ -0,0 +1,76 @@
+    FutureSet futures;
+    futures.add(AddTask(*threadPool, failingTask<TestException1>()));
+    futures.add(AddTask(*threadPool, failingTask<TestException2>()));
+    BOOST_CHECK_THROW(futures.join(), TestException1);
```
Is the task created at l. 73 guaranteed to be executed before the one at l. 74 ?

If there is no such guarantee, we can only expect that an exception among `TestException1` and `TestException2` will be thrown & caught, but not necessarily the 1st one.

sylvlecl replied:
The execution order is clearly not guaranteed, but the exception thrown will be the one corresponding to the first **submitted** task that fails, not the first exception that occurs.

flomnes commented on src/libs/antares/concurrency/include/antares/concurrency/concurrency.h:
```diff
@@ -0,0 +1,49 @@
+{
+
+typedef std::function<void()> Task;
+typedef std::future<void> TaskFuture;
```
```suggestion
using Task = std::function<void()>;
using TaskFuture = std::future<void>;
```

flomnes commented on src/libs/antares/concurrency/concurrency.cpp:
```diff
@@ -0,0 +1,53 @@
+
+std::unique_ptr<Yuni::Job::IJob> MakeJob(const Task& task) {
+    return std::make_unique<JobImpl>(task);
+}
```
unused

flomnes commented on src/libs/antares/concurrency/concurrency.cpp:
```diff
@@ -0,0 +1,53 @@
+
+std::unique_ptr<JobImpl> MakePackagedJob(const Task& task) {
+    return std::make_unique<JobImpl>(task);
+}
```
Only used internally, make static or place in anonymous NS

Note that in C++20 `auto` [is a valid type](https://stackoverflow.com/a/60355539) for methods.

flomnes commented on src/libs/antares/concurrency/concurrency.cpp:
```diff
@@ -0,0 +1,53 @@
+std::future<void> AddTask(Yuni::Job::QueueService& threadPool, const Task& task) {
+    auto job = MakePackagedJob(task);
+    auto future = job->getFuture();
+    threadPool.add(Yuni::Job::IJob::Ptr(job.release()));
```
Does this compile ?
```suggestion
    threadPool.add(job.release());
```

sylvlecl committed:
- transfer exceptions occuring in zip_writer too
- some cleanup
- an integration test to check behaviour in simulation

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

flomnes commented on src/libs/antares/concurrency/concurrency.cpp:
```diff
@@ -0,0 +1,53 @@
+}
+
+void FutureSet::join() {
+    for (auto& f: futures_) {
+        f.get();
+    }
```
This might require a `mutex` to prevent concurrent access

sylvlecl replied:
After more thoughts, I actually added a mutex to make all methods thread safe, in particular the addition of new futures, since vector are not thread safe.

flomnes commented on src/libs/antares/concurrency/concurrency.cpp:
```diff
@@ -0,0 +1,56 @@
+std::future<void> AddTask(Yuni::Job::QueueService& threadPool,
+                          const Task& task,
+                          Yuni::Job::Priority priority) {
+    auto job = std::make_unique<PackagedJob>(task);
+    auto future = job->getFuture();
+    threadPool.add(Yuni::Job::IJob::Ptr(job.release()), priority);
```
Being the devil's advocate, why bother using a `std::unique_ptr` if you `release` it 2 lines after ?

```suggestion
    PackageJob* job = new PackagedJob(task);
    auto future = job->getFuture();
    threadPool.add(Yuni::Job::IJob::Ptr(job), priority);
```

flomnes commented on src/solver/simulation/solver.hxx:
```diff
@@ -45,11 +45,15 @@
 namespace Antares::Solver::Simulation
 {
 
+using namespace Antares::Concurrency;
```
You could be a bit more specific and name what classes/function you'll be using

flomnes commented on src/libs/antares/writer/zip_writer.cpp:
```diff
@@ -46,7 +46,7 @@ static std::unique_ptr<mz_zip_file> createInfo(const std::string& entryPath)
 
 template<class ContentT>
-void ZipWriteJob<ContentT>::onExecute()
+void ZipWriteJob<ContentT>::writeEntry()
```
I see a few `logs.error()` below. Could we also `throw` exceptions now ?

sylvlecl replied:
Indeed: it will change a little the behaviour from what it is now, but I think it's better to get an error instead of having an incomplete/corrupt result without knowing it (who looks at the log? :) ).

flomnes commented on src/libs/antares/writer/private/zip_writer.hxx:
```diff
@@ -11,8 +12,9 @@ void ZipWriter::addEntryFromBufferHelper(const std::string& entryPath, ContentTy
-    pQueueService->add(
-      new ZipWriteJob<ContentType>(*this, entryPath, entryContent, pDurationCollector),
-      Yuni::Job::priorityLow);
+    Antares::Concurrency::AddTask(*pQueueService,
+                                  ZipWriteJob<ContentType>(*this, entryPath, entryContent, pDurationCollector),
+                                  Yuni::Job::priorityLow);
```
Unfortunately, the result of this is not stored in a `FutureSet`, so we won't be able to catch exceptions.

Maybe we could wrep `QueueService` and `FutureSet` together and expose a simpler `add` method ?

```cpp
class Wrapper {
Yuni::QueueService* queue;
FutureSet results;
public:
add(...);
};
```
Note that `Wrapper` may be returned by `createThreadPool` that is defined in the tests, instead of a mere QueueService.

sylvlecl replied:
Indeed the simple call to `AddTask` is not useful by itself.

What I did to cope with it is:
- Addition of [[nodiscard]] to the method to be warned when we ignore the future
- Addition of a `flush()` method to the writer, in order to be able to synchronize a thread with underlying I/O tasks, and get possible exception from here

The writer does not use the same thread pool as computations any more, but its own one, which is never started/stopped.
The only synchonization mechanism used for I/O now the flush method.

sylvlecl committed:
Take into account review

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
- Separate thread pool for zip IO
- Addition of a flush method for writers, to synchronize
- Addition of tests for zip writer

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Some comments

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

flomnes commented on src/libs/antares/concurrency/concurrency.cpp:
```diff
@@ -0,0 +1,61 @@
+    {
+        std::lock_guard lock(mutex_);
+        moved = std::move(futures_);
+        futures_ = std::vector<TaskFuture>{};
```
```suggestion
        std::swap(moved, futures_);
```

sylvlecl replied:
Done indeed!

sylvlecl committed:
Throw more exceptions from zip writer

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix headers

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix missing header (windows build)

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Merge remote-tracking branch 'origin/develop' into fix/fix-silent-exceptions

flomnes commented on src/libs/antares/writer/zip_writer.cpp:
```diff
@@ -82,12 +96,14 @@ ZipWriter::ZipWriter(std::shared_ptr<Yuni::Job::QueueService> qs,
  pArchivePath(std::string(archivePath) + ".zip"),
- pDurationCollector(duration_collector)
+ pDurationCollector(duration_collector),
+ pendingTasks_()
```
Calling the default constructor is not necessary, it will be called implicitly

sylvlecl replied:
Removed

flomnes commented on src/libs/antares/writer/zip_writer.cpp:
```diff
@@ -82,12 +96,14 @@ ZipWriter::ZipWriter(std::shared_ptr<Yuni::Job::QueueService> qs,
     if (int32_t ret = mz_zip_writer_open_file(pZipHandle, pArchivePath.c_str(), 0, 0); ret != MZ_OK)
     {
         logs.error() << "Error opening zip file " << pArchivePath << " (" << ret << ")";
+        logErrorAndThrow("Error opening zip file " + pArchivePath + " (" + std::to_string(ret) + ")");
```
Please remove the previous line (redundant)

sylvlecl replied:
Removed

flomnes commented on src/tests/src/libs/antares/writer/test_zip_writer.cpp:
A nice addition !

sylvlecl replied:
It's a little clumsy though, with the minizip API !
If we use more zip in the future, it would be nice to add a small c++ wrapper around that C API :)

Also I did not really find a way to cause a failure in the writing of entries ... if you have any idea about this it's welcome.

flomnes replied:
Providing negative `len` maybe ? Or an invalid path ?

sylvlecl committed:
Take into account review

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix windows build again

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

flomnes commented on src/tests/src/libs/antares/writer/test_zip_writer.cpp:
```diff
@@ -0,0 +1,106 @@
+using Antares::Solver::IResultWriter;
+
+// Handles lifetime of necessary objects
+struct TestContext
+{
+    std::shared_ptr<QueueService> threadPool;
+    std::unique_ptr<IDurationCollector> durationCollector;
+    std::shared_ptr<IResultWriter> writer;
+};
```
Fixture ?

sylvlecl replied:
I did not investigate it so much, but I cannot easily find a good use here: 
to create those objects I need to pass some parameters (path, threads). Fixtures are limited to only one constructor argument. It could be a small struct but it becomes weird...

Those arguments are not strictly necessary, but would be useful for additional tests.

sylvlecl commented on src/solver/application.cpp:
```diff
@@ -281,8 +281,11 @@ void Application::processCaption(const Yuni::String& caption)
 void Application::prepareWriter(Antares::Data::Study& study,
                                 Benchmarking::IDurationCollector& duration_collector)
 {
+    ioQueueService = std::make_shared<Yuni::Job::QueueService>();
```
@flomnes : Note that the zip writing is now handled by this separate thread pool which is allocated only 1 thread.

Since I don't do "introspection" on the type of writer used, the thread is created also when using plain files, and not used, which was not the case before.
The runtime negligible cost seems reasonable to me.

flomnes replied:
You could use `IResultWriter::needsTheJobQueue()`, or even make `EnsureQueueStartedIfNeeded` public and use it.

sylvlecl replied:
The writer is not yet created when we need to know this:
we can instead have a factory interface (`ResultWriterFactory` with 2 implementations) which provides that information before creation of the writer, but honestly I think this is overkill, having a sleeping thread is really not an issue.

It would make responsibilities clearer to leave it as is: the application provides a usable thread pool, the writer has the choice to use it or not.
Actually, I would also be in favor of removing `EnsureQueueStartedIfNeeded` completely, together with the `needsTheJobQueue()` in the interface, to make things simpler. 
It's the responsibility of the application to provide a working thread pool, the writer should not need to worry about starting anything.

sylvlecl replied:
Or alternatively the writer could maintain its own working thread, but it would make it quite "hidden".

flomnes replied:
Ok, let us keep it like that

a-zakir commented on src/libs/antares/concurrency/include/antares/concurrency/concurrency.h:
```diff
@@ -0,0 +1,88 @@
+
+/*!
+ * \brief Utility class to gather futures to wait for.
+ */
```
yeah pretty doc ;)
